docs/declarative.md

```diff
@@ -0,0 +1,507 @@
+  - Declarative HTML tools present in the current DOM (elements with `tool-name`).
+  - JavaScript tools registered via `window.mcp.registerTool(...)` in that browsing context.
+- Tool names MUST be unique within a catalog. If a name collision occurs, the browser SHOULD prefer the most recently registered definition and MAY surface a warning in DevTools.
+
+>The MCP spec throws and error on tool name collision. We should probably do the same. Tool collisions are far more likely to happen in WebMCP servers as there is now up to three possible sources of tools: declarative, javascript, and service worker.
+
+Change detection and ordering guarantees
+- Tools can appear or disappear as a result of a tool call (e.g., DOM updates, conditional registration).
+- To preserve correctness and avoid race conditions, the browser MUST enforce this ordering for a single tool invocation:
+  1) Execute the tool and fully evaluate its body (JS `execute` or form submission/response parsing).
+  2) Deliver the tool result to the client (agent) over MCP.
+  3) Recompute the catalog (scan DOM + JS registrations), compute deltas, and if changed, emit [`notifications/tools/list_changed`](https://modelcontextprotocol.io/specification/2025-06-18/schema#notifications%2Ftools%2Flist-changed) to all connected clients for this server.
+  4) Apply any navigation/unmounts/redirects (e.g., those implied by `_meta.uiRedirect` or HTTP redirects).
+- In particular, the tool that is currently executing MUST NOT be unmounted (e.g., DOM removal or JS unregistration) until after its result has been delivered to the client and any [`notifications/tools/list_changed`](https://modelcontextprotocol.io/specification/2025-06-18/schema#notifications%2Ftools%2Flist-changed) notification has been sent.
+
+Notification shape
+- Method: [`notifications/tools/list_changed`](https://modelcontextprotocol.io/specification/2025-06-18/schema#notifications%2Ftools%2Flist-changed)
+- Use when: any change to the catalog (addition, removal, or metadata/schema change) is observed after a recomputation.
+
+```
+interface ToolListChangedNotification {
+  method: "notifications/tools/list_changed";
+  params?: { _meta?: { [key: string]: unknown }; [key: string]: unknown };
+}
+```
+
+Reference: MCP notification — notifications/tools/list_changed: https://modelcontextprotocol.io/specification/2025-06-18/schema#notifications%2Ftools%2Flist-changed
+
+Sequence (tool call with catalog update and redirect)
+
+```mermaid
+sequenceDiagram
+participant A as Agent (client)
+participant B as Browser (page server)
+participant P as Page (DOM/JS)
+participant S as Backend (optional)
+
+A->>B: Call tool T(params)
+alt JS tool
+  B->>P: execute(params)
+  P-->>B: result
+else Declarative form
+  B->>S: fetch(action, method, body/query) [Accept: JSON]
+  S-->>B: JSON result (optional Location)
+end
+B-->>A: Return CallToolResult
+B->>P: Recompute catalog (DOM + JS)
```

Comment:
I think this step is what I was commenting on earlier. I'm not totally sure why we'd need to recompute the list of available tools at this stage. I'd assumed this would be automatically kept up to date by operations that add/remove declarative tools (or imperative, JS-based tools) on its own. Is it possible that the `CallToolResult` JSON specifies new tools to be added?